drivers/mspi/mspi_cadence.c

```diff
@@ -0,0 +1,859 @@
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, PERIPH_SEL_DEC, base_addr);
+
+	/* CS selection is based on "manual" pin selection */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, ENABLE_AHB_DECODER, base_addr);
+
+	/* DQ3 should not be used as reset pin */
+	MSPI_CADENCE_REG_WRITE(1, CONFIG, RESET_CFG, base_addr);
+
+	/* Set baud rate division to 32; formula: (n + 1) * 2 */
+	MSPI_CADENCE_REG_WRITE(15, CONFIG, MSTR_BAUD_DIV, base_addr);
+
+	/* disable dual byte opcodes */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, DUAL_BYTE_OPCODE_EN, base_addr);
+
+	/* disable PHY pipeline mode */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, PIPELINE_PHY, base_addr);
+
+	/* disable PHY module generally */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, PHY_MODE_ENABLE, base_addr);
+
+	/* disable CRC */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, CRC_ENABLE, base_addr);
+
+	/* disable DMA generally since it's not supported */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, ENB_DMA_IF, base_addr);
+
+	/* disable automatic write protection disablement of MSPI peripherals */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, WR_PROT_FLASH, base_addr);
+
+	/* disable possible reset pin */
+	MSPI_CADENCE_REG_WRITE(0, CONFIG, RESET_PIN, base_addr);
+
+	/* general clock cycle delays */
+	MSPI_CADENCE_REG_WRITE(timing_config->nss, DEV_DELAY, D_NSS, base_addr);
+	MSPI_CADENCE_REG_WRITE(timing_config->btwn, DEV_DELAY, D_BTWN, base_addr);
+	MSPI_CADENCE_REG_WRITE(timing_config->after, DEV_DELAY, D_AFTER, base_addr);
+	MSPI_CADENCE_REG_WRITE(timing_config->init, DEV_DELAY, D_INIT, base_addr);
+
+	/* set trigger reg address and range to 0 */
+	MSPI_CADENCE_REG_WRITE(0, IND_AHB_ADDR_TRIGGER, ADDR, base_addr);
+	MSPI_CADENCE_REG_WRITE(0, INDIRECT_TRIGGER_ADDR_RANGE, IND_RANGE_WIDTH, base_addr);
+
+	/* disable loop-back via DQS */
+	MSPI_CADENCE_REG_WRITE(1, RD_DATA_CAPTURE, BYPASS, base_addr);
+
+	/* disable auto polling for write completion */
+	MSPI_CADENCE_REG_WRITE(1, WRITE_COMPLETION_CTRL, DISABLE_POLLING, base_addr);
+
+	/* disable automatic write enable command before indirect write transactions */
+	MSPI_CADENCE_REG_WRITE(0, DEV_INSTR_WR_CONFIG, WEL_DIS, base_addr);
+
+	/* reset mode bit (hardware CRC checking on read, if supported) */
+	MSPI_CADENCE_REG_WRITE(0, DEV_INSTR_RD_CONFIG, MODE_BIT_ENABLE, base_addr);
+
+	/* disable DDR mode */
+	MSPI_CADENCE_REG_WRITE(0, DEV_INSTR_RD_CONFIG, DDR_EN, base_addr);
+
+	uint32_t val;
+
+	/* disable all interrupts via masking */
+	val = sys_read32(base_addr + CADENCE_MSPI_IRQ_MASK_REG);
+	val &= ~CADENCE_MSPI_IRQ_MASK_ALL;
+	sys_write32(val, base_addr + CADENCE_MSPI_IRQ_MASK_REG);
+
+	/* clear currently pending interrupts */
+	val = sys_read32(base_addr + CADENCE_MSPI_IRQ_STATUS_REG);
+	val |= CADENCE_MSPI_IRQ_STATUS_ALL;
+	sys_write32(val, base_addr + CADENCE_MSPI_IRQ_STATUS_REG);
+
+	/* re-enable MSPI controller */
+	MSPI_CADENCE_REG_WRITE(1, CONFIG, ENABLE_SPI, base_addr);
+
+	return 0;
+}
+
+static int mspi_cadence_small_transfer(const struct device *controller, const struct mspi_xfer *req,
```

Comment:
minor suggestion: I think `mspi_cadence_stig` is a more suitable name for this function